src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +153,137 @@ private void logTargetPath() {
+        final AprilTagCamera[] newResultCameras = getCamerasWithNewResult();
+        sortCamerasByLatestResultTimestamp(newResultCameras);
+
+        updateEstimatedPoseFromVision(newResultCameras);
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private AprilTagCamera[] getCamerasWithNewResult() {
+        final AprilTagCamera[] camerasWithNewResult = new AprilTagCamera[aprilTagCameras.length];
+        int index = 0;
+
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            aprilTagCamera.update();
+            if (aprilTagCamera.hasNewResult()) {
+                camerasWithNewResult[index] = aprilTagCamera;
+                index++;
+            }
         }
-        return viableVisionObservations;
+
+        return Arrays.copyOf(camerasWithNewResult, index);
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
-        final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
-        if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
-            return null;
+    private void sortCamerasByLatestResultTimestamp(AprilTagCamera[] aprilTagCameras) {
+        QuickSort.sort(aprilTagCameras, AprilTagCamera::getLatestResultTimestampSeconds);
+    }
 
-        return new PoseEstimator6328.VisionObservation(
-                robotPose,
-                aprilTagCamera.getLatestResultTimestampSeconds(),
-                aprilTagCamera.calculateStandardDeviations()
-        );
+    /**
+     * Updates the estimated pose from each camera, each at the timestamp of their latest result.
+     */
+    private void updateEstimatedPoseFromVision(AprilTagCamera[] aprilTagCameras) {
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
+            addVisionObservation(
+                    aprilTagCamera.getEstimatedRobotPose(),
+                    aprilTagCamera.calculateStandardDeviations(),
+                    aprilTagCamera.getLatestResultTimestampSeconds()
+            );
+        }
     }
 
-    private void putAprilTagsOnFieldWidget() {
-        for (Map.Entry<Integer, Pose3d> entry : FieldConstants.TAG_ID_TO_POSE.entrySet()) {
-            final Pose2d tagPose = entry.getValue().toPose2d();
-            field.getObject("Tag " + entry.getKey()).setPose(tagPose);
+    private void addVisionObservation(Pose2d estimatedPose, StandardDeviations standardDeviations, double timestamp) {
```

Comment:
In this function and its predecessors, it's not clear which is the estimator's estimate and which is the vision estimate, makes everything really confusing